kotlin-coroutines-end/app/src/main/java/com/example/android/kotlincoroutines/main/RefreshMainDataWork.kt

```diff
@@ -18,44 +18,40 @@ package com.example.android.kotlincoroutines.main
 
 import android.content.Context
 import androidx.annotation.WorkerThread
+import androidx.work.CoroutineWorker
 import androidx.work.Worker
 import androidx.work.WorkerParameters
-import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.*
 
 /**
- * Worker job to refresh refresh titles from the network while the app is in the background.
+ * Worker job to refresh titles from the network while the app is in the background.
  *
  * WorkManager is a library used to enqueue work that is guaranteed to execute after its constraints
  * are met. It can run work even when the app is in the background, or not running.
  */
-class RefreshMainDataWork(context: Context, params: WorkerParameters) : Worker(context, params) {
+class RefreshMainDataWork(context: Context, params: WorkerParameters) :
+        CoroutineWorker(context, params) {
+
+    // CoroutineWorker defaults to Dispatchers.Default
+    // We want to execute this on Dispatchers.IO
+    override val coroutineContext = Dispatchers.IO
```

Comment:
This isn't required since the repository already does thread confinement. Switching from Default to IO is not a thread switch.

Better to avoid showing this since it will lead people to thinking they should perform thread confinement at this level.

Reply:
Also, TODO for me or @manuelvicnt we should add some more details about thread confinement as a concept to the text of the codelab since it's a hero use case on Android.

Reply:
This code is not guaranteed to be executed in the repository. This is deferred work that can be launched at any point in the future, e.g. after getting Wi-Fi connection or after reboot. Defining `coroutineContext` is needed here.

Reply:
Hm, I'm not following. This code will be run on a background thread pool managed by WorkManager, by default the CoroutineWorker will start the job in Dispatchers.Default, which uses the same thread pool as Dispatchers.IO. Switching between these Dispatchers never causes a thread switch (but does modify some tracking counters). What problem is this trying to solve?

Reply:
I don't believe this code is needed. It does make sense if this job made a lot of calls to functions that were confined on an Executor based dispatcher which would cause this optimal code path to run ([equals](https://github.com/Kotlin/kotlinx.coroutines/blob/master/core/kotlinx-coroutines-core/src/Executors.kt#L114), withContext [fast path 2](https://github.com/Kotlin/kotlinx.coroutines/blob/master/common/kotlinx-coroutines-core-common/src/Builders.common.kt#L140))

In our case, this will save a switch from Default to IO which does save an object allocation and some checks, but it does not save a thread switch so the overall cost here is pretty small.

Note the comment on [Dispatchers.IO](https://github.com/Kotlin/kotlinx.coroutines/blob/master/core/kotlinx-coroutines-core/src/Dispatchers.kt#L77) talking about thread switch behavior between these dispatchers. WorkManager will start our jobs on Dispatchers.Default for us.

Reply:
doWork calls `repository.refreshTitle()`, a suspended function that it's using a `withContext(Dispatchers.IO)` to run it's "networking code".

I can remove this `override` and have `doWork()` launched on `Dispatcher.Default`, but we're going to land on `Dispatcher.IO` sooner or later.

Reply:
'doWork()' is now run on `Dispatcher.Default`.